queasars/job_shop_scheduling/domain_wall_hamiltonian_encoder.py

```diff
@@ -0,0 +1,528 @@
+        For the reasoning behind this see: https://iopscience.iop.org/article/10.1088/2058-9565/ab33c2/meta
+        """
+        if i < -1 or i > self.n_qubits:
+            raise ValueError("The index is out of the bounds of the domain wall variable!")
+        if i == -1:
+            return -1 * _constant_one_term(n_qubits=quantum_circuit_n_qubits)
+        if i == self.n_qubits:
+            return _constant_one_term(n_qubits=quantum_circuit_n_qubits)
+        return _pauli_z_term(qubit_index=self._qubit_start_index + i, n_qubits=quantum_circuit_n_qubits)
+
+    @property
+    def values(self) -> tuple[int, ...]:
+        """
+        :return: the values between which this domain wall variable chooses
+        :rtype: tuple[int, ...]
+        """
+        return self._values
+
+    @property
+    def max_value(self) -> float:
+        """
+        :return: the maximum value contained in the domain wall variable's values
+        :rtype: int
+        """
+        return self._max_value
+
+    @property
+    def min_value(self) -> float:
+        """
+        :return: the minimum value contained in the domain wall variable's values
+        :rtype: int
+        """
+        return self._min_value
+
+    @property
+    def n_qubits(self) -> int:
+        """
+        :return: the amount of qubits needed by this variable
+        """
+        return self._n_qubits
+
+    def viability_term(self, penalty: float, quantum_circuit_n_qubits: int) -> SparsePauliOp:
+        """
+        Returns a SparsepauliOp which penalizes invalid variable states (states with more than one domain
+        wall). Within a hamiltonian this term evaluates to 0 only if the variable is in a valid state and to
+        (n-1)*penalty at maximum for the n values this variable can represent
+
+        :arg penalty: size of the applied penalty for each violation
+        :type penalty: float
+        :arg quantum_circuit_n_qubits: the amount of qubits in the quantum circuit in which this variable is part of
+        :type quantum_circuit_n_qubits: int
+        :return: a SparsePauliOp which penalizes invalid variable states
+        :rtype: SparsePauliOp
+        """
+        if self._n_qubits == 0:
+            return 0 * _constant_one_term(n_qubits=quantum_circuit_n_qubits)
+
+        penalty = penalty / 2
+        local_terms: list[SparsePauliOp] = []
+        for i in range(-1, self._n_qubits):
+            local_terms.append(
+                1
+                / 2
+                * penalty
+                * (
+                    _constant_one_term(n_qubits=quantum_circuit_n_qubits)
+                    - self._z_dash_term(
+                        i=i,
+                        quantum_circuit_n_qubits=quantum_circuit_n_qubits,
+                    ).compose(
+                        self._z_dash_term(
+                            i=i + 1,
+                            quantum_circuit_n_qubits=quantum_circuit_n_qubits,
```

Comment:
This is intended here. For details see the section 2 of the [domain wall paper](https://iopscience.iop.org/article/10.1088/2058-9565/ab33c2/meta). I will gladly explain this in person, it would just be a little much to write here.

Reply:
Ok, I see, thanks!